### PR TITLE
Fix certificate rotation problem

### PR DESCRIPTION
**Pull request by lei-tang**
Please provide a description for what this PR is for: https://github.com/istio/istio/issues/21239

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: security/pkg/nodeagent/cache/secretcache.go [hunk 1]
```diff
@@ -551,6 +551,12 @@ func (sc *SecretCache) rotate(updateRootFlag bool) {
 			atomic.AddUint64(&sc.secretChangedCount, 1)
 
 			// Send the notification to close the stream if token is expired, so that client could re-connect with a new token.
+			// In its current implementation, isTokenExpired() does not really check whether
+			// a token expired or not and always returns true.
+			// When isTokenExpired() returns true, the stream connection to Envoy will
+			// be closed and Envoy will reconnect with a token.
+			// There are multiple ways to obtain a token, e.g., read from a file,
+			// and call a token API, which is up to the Envoy config.
 			if sc.isTokenExpired() {
```
Comment (howardjohn) [inline security/pkg/nodeagent/cache/secretcache.go]: Can we fix this code? It can be in another PR, but closing the connection isn't really a good solution

File: pkg/istio-agent/sds-agent.go [hunk 2]
```diff
@@ -430,6 +430,9 @@ func applyEnvVars() {
 	workloadSdsCacheOptions.SecretTTL = secretTTLEnv
 	workloadSdsCacheOptions.SecretRefreshGraceDuration = secretRefreshGraceDurationEnv
 	workloadSdsCacheOptions.RotationInterval = secretRotationIntervalEnv
+	// If the secret is not rotated for 3*workloadSdsCacheOptions.SecretTTL,
+	// it is regarded as stale and will be evicted.
+	workloadSdsCacheOptions.EvictionDuration = 3 * workloadSdsCacheOptions.SecretTTL
```
Comment (howardjohn) [inline pkg/istio-agent/sds-agent.go]: This option is pretty scary... basically if we ever miss a cert rotation for whatever reason and we end up hitting this when we shouldn't, we are going to be permanently broken and never rotate certs? Why do we even need this for workload, there is only even `default` and `ROOTCA`?

File: pkg/istio-agent/sds-agent.go [hunk 3]
```diff
@@ -430,6 +430,9 @@ func applyEnvVars() {
 	workloadSdsCacheOptions.SecretTTL = secretTTLEnv
```
Comment (howardjohn) [inline pkg/istio-agent/sds-agent.go]: Can we also default the grace period and refresh interval based on this one? Otherwise its so easy to shoot yourself in the foot if you don't configure all 3 options

File: pkg/istio-agent/sds-agent.go (hunk 2 above)
Comment (myidpt) [reply pkg/istio-agent/sds-agent.go]: I don't think we need this eviction functionality for istio agent, since unlike node agent, it's only serving a single sidecar.

File: security/pkg/nodeagent/cache/secretcache.go (hunk 1 above)
Comment (myidpt) [reply security/pkg/nodeagent/cache/secretcache.go]: Yes, we should fix it in another PR.

**Comment by myidpt** [comment]
This is a better fix: https://github.com/istio/istio/pull/21274
Not bothering Lei since he's in different time zone :)
